File: CrowdAssist/app/src/main/java/th/ac/kmitl/it/crowdalert/SignUpActivity.java
package th.ac.kmitl.it.crowdalert;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import th.ac.kmitl.it.crowdalert.component.TermsAndConditionsDialog;

public class SignUpActivity extends AppCompatActivity implements View.OnClickListener {
    private FirebaseAuth mAuth;
    private FirebaseDatabase mDatabase;
    private EditText username;
    private EditText password;
    private EditText repeatPassword;
    private EditText email;
    private EditText firstName;
    private EditText lastName;
    private EditText idCard;
    private LinearLayout signUp;
    private Boolean firstTime = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sign_up);
        this.getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN );
        mAuth = FirebaseAuth.getInstance();
        mDatabase = FirebaseDatabase.getInstance();
        username = findViewById(R.id.inputUsername);
        password = findViewById(R.id.inputPassword);
        repeatPassword = findViewById(R.id.inputConfirmPassword);
        email = findViewById(R.id.inputEmail);
        firstName = findViewById(R.id.inputFirstname);
        lastName = findViewById(R.id.inputLastname);
        idCard = findViewById(R.id.inputIDcard);
        signUp = findViewById(R.id.sign_up_button);

        signUp.setOnClickListener(this);
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (firstTime){
            TermsAndConditionsDialog dialog = new TermsAndConditionsDialog(this);
            dialog.show();
            firstTime = false;
        }
    }

    @Override
    public void onClick(View view) {
        if (!password.getText().toString().equals(repeatPassword.getText().toString())){
            return;
        }
        mDatabase.getReference("user_id").child(username.getText().toString()).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                if (!dataSnapshot.exists()){
                    mAuth.createUserWithEmailAndPassword(email.getText().toString(), password.getText().toString())
                            .addOnCompleteListener(SignUpActivity.this, new OnCompleteListener<AuthResult>() {
                                @Override
                                public void onComplete(@NonNull Task<AuthResult> task) {
                                    if (task.isComplete()){
                                        FirebaseUser user = mAuth.getCurrentUser();
                                        UserSignUp information = new UserSignUp(username.getText().toString(), email.getText().toString(), firstName.getText().toString(), lastName.getText().toString(), idCard.getText().toString(), "user");
                                        mDatabase.getReference("users").child(user.getUid()).setValue(information);
                                        mDatabase.getReference("user_id").child(username.getText().toString()).setValue(email.getText().toString());
                                        finish();
                                    }else{

                                    }
                                }
                            });
                }else{
                    Toast.makeText(SignUpActivity.this, "Username นี้มีคนใช้งานแล้ว", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });
    }

    public static class UserSignUp{
        public String username;
        public String email;
        public String firstName;
        public String lastName;
        public String idCard;
        public String role;
        public Boolean firstTime;
        public Boolean verify;

        public UserSignUp() {
        }

        public UserSignUp(String username, String email, String firstName, String lastName, String idCard, String role) {
            this.username = username;
            this.email = email;
            this.firstName = firstName;
            this.lastName = lastName;
            this.idCard = idCard;
            this.role = role;
            this.firstTime = true;
            this.verify = false;
        }
    }
}

File: CrowdAssist/app/src/main/java/th/ac/kmitl/it/crowdassist/SignUpActivity.kt
package th.ac.kmitl.it.crowdassist

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.constraint.ConstraintLayout
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.EditText
import android.widget.LinearLayout
import th.ac.kmitl.it.crowdassist.component.TermsAndConditionsDialog
import th.ac.kmitl.it.crowdassist.contract.SignUpContract
import th.ac.kmitl.it.crowdassist.presenter.SignUpPresenter
import th.ac.kmitl.it.crowdassist.util.DatabaseHelper
import kotlin.properties.Delegates

class SignUpActivity : AppCompatActivity(), SignUpContract.View{
    var termsAndConditionDialog : TermsAndConditionsDialog by Delegates.notNull<TermsAndConditionsDialog>()
    var firstTime = true
    val presenter = SignUpPresenter(DatabaseHelper(this), this)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sign_up)
        termsAndConditionDialog = TermsAndConditionsDialog(this)
        val signUpButton = findViewById<LinearLayout>(R.id.sign_up_button)
        signUpButton.setOnClickListener({
            presenter.onSignUpClicked()
        })
    }

    override fun showProgressBar() {
        findViewById<ConstraintLayout>(R.id.mainLayout).visibility = View.GONE
        findViewById<LinearLayout>(R.id.progress_layout).visibility = View.VISIBLE
    }

    override fun hideProgressBar() {
        findViewById<ConstraintLayout>(R.id.mainLayout).visibility = View.VISIBLE
        findViewById<LinearLayout>(R.id.progress_layout).visibility = View.GONE
    }

    override fun showTermsAndConditions() {
        termsAndConditionDialog.show()
    }

    override fun getAllTextFill(): MutableMap<String, String>? {
        val textFills = hashMapOf<String, String>()
        textFills.put("username", (findViewById<EditText>(R.id.username_input)).text.toString())
        textFills.put("password", (findViewById<EditText>(R.id.password_input)).text.toString())
        textFills.put("rePassword", (findViewById<EditText>(R.id.confirm_password_input)).text.toString())
        textFills.put("firstName", (findViewById<EditText>(R.id.firstname_input)).text.toString())
        textFills.put("lastName", (findViewById<EditText>(R.id.lastname_input)).text.toString())
        textFills.put("email", (findViewById<EditText>(R.id.email_input)).text.toString())
        textFills.put("idCard", (findViewById<EditText>(R.id.idcard_input)).text.toString())
        return textFills
    }

    override fun showSnackBar(message: String, during: Int) {
        Snackbar.make(findViewById<ConstraintLayout>(R.id.mainLayout), message , during).show()
    }

    override fun onStart() {
        super.onStart()
        if (firstTime) {
            showTermsAndConditions()
            firstTime = false
        }
    }

    override fun startActivity(activity : Class<out Activity>){
        val intent = Intent(this, activity)
        startActivity(intent)
    }
}
